### ISSUE TITLE
Provide guidance and error type for integrations raising errors

### ISSUE THREAD

**Issue by aiwilliams** [JupiterOne/integration-sdk#156]
The gitlab integration stores an [API response code in `IntegrationError.code`](https://github.com/JupiterOne/graph-gitlab/blob/3c8fd0f19f4a16915788f0338926e0d2c10b2187/src/provider/GitlabClient.ts#L131), and there may be another integration doing something similar. However, I think now that was not the best choice.

In https://github.com//JupiterOne/integration-sdk/blob/master/src/errors.ts#L7, and everywhere in the integration-sdk currently, the values for `code` are constant-like strings, and there are error classes that assign `code` so that we don't see the strings repeated where the errors are thrown (great!).

We should avoid having each integration develop their own set of codes for common things, wo that we can make some things consistent for runtime monitoring and debugging. We'll need a clear way for integrations to raise certain types of errors, and limit the use of new error codes as much as reasonable.

Consider this proposal:

```ts
/**
 * A type representing errors generated or handled by the integration runtime.
 * Integrations should not throw this error type directly. See below for
 * options.
 *
 * All errors occurring during execution of an integration should ultimately end
 * up as an `IntegrationError`, or a subclass thereof, each expressing a `code`
 * reflecting the class of the error. In case an unhandled `Error` occurs, which
 * has no `code`, then `"UNEXPECTED_ERROR"` will be used in logging the error.
 *
 * Integration code should throw these exceptions:
 *
 * ```
 *   // Thrown during `validateInvocation`, anything the user should see that
 *   // can help them fix the problem on their own.
 *   throw new IntegrationValidationError("Something that will be displayed to user so they can fix configuration")
 *
 *   // Typically thrown during `validateInvocation`. User will be informed of
 *   // what we're calling to test authentication, and the response code.
 *   throw new IntegrationProviderAuthenticationError({
 *     cause: err,
 *     endpoint: "https://something.com/api/authcheckwhatever",
 *     status?: err.status,
 *     message?: err.statusText
 *   });
 *
 *   // Thrown during resource fetching steps. The runtime will show these along
 *   // with information about what could not be obtained thanks to insufficient
 *   // access.
 *   throw new IntegrationProviderAuthorizationError({
 *     type: [], // the `_type`s that could not be obtained
 *     cause: err,
 *     endpoint: "https://something.com/api/resource",
 *     status?: err.status,
 *     message?: err.statusText
 *   });
 *
 *   // Thrown for unexpected API errors. This will be shown to users to help
 *   // them understand why some data sets may be incomplete.
 *   throw new IntegrationProviderAPIError({
 *     cause: err,
 *     endpoint: "https://something.com/api/something",
 *     status?: err.status,
 *     message?: err.statusText
 *   });
 * ```
 */
export class IntegrationError extends Error { ... }
```

**Comment by charlieduong94** [comment]
I like this proposal. I imagine that we will have something like `logger.authError` that will work with these errors to provide a clear message that some action needs to be taken.

**Comment by aiwilliams** [comment]
I believe this is closed by #157 , and further work is represented by #136 .
